doc/example-ast-destructing.mld

```diff
@@ -0,0 +1,171 @@
+
+Let’s say we want to destructure an AST representing the integer [1]:
+
+{[
+let match_int_payload ~loc payload =
+  match payload with
+  | PStr
+      [
+        {
+          pstr_desc =
+            Pstr_eval
+              ({ pexp_desc = Pexp_constant (Pconst_integer (value, None)); _ }, _);
+          _;
+        };
+      ] -> (
+      try Ok (value |> int_of_string)
+      with Failure _ ->
+        Error (Location.Error.createf ~loc "Value is not a valid integer"))
+  | _ -> Error (Location.Error.createf ~loc "Wrong pattern")
+]}
+
+1. {b Pattern Matching the Payload}:
+    - Begins by matching the [payload] with the expected structure.
+    - The pattern expects a structure ([PStr]) containing a single item.
+2. {b Destructuring the Structure Item}:
+    - Matches the [pstr_desc] field, expecting an evaluated expression ([Pstr_eval]).
+    - The expression should be a constant integer ([Pexp_constant] with [Pconst_integer]).
+    - Captures the integer value as a string in [value].
+3. {b Handling the Matched Value}:
+    - Converts the [value] to an integer and returns [Ok] if successful.
+    - If conversion fails, returns an error message.
+4. {b Handling Mismatched Patterns}:
+    - If the [payload] doesn’t match the expected structure, it returns an error.
+
+While this method is powerful, it can be verbose and difficult to maintain as patterns become more complex.
+
+{1:using-ast_pattern-high-level-destructors Using {!Ppxlib.Ast_pattern} High-Level Destructors}
+
+To make AST destructuring more readable, PPXLib provides the {!Ppxlib.Ast_pattern} module, which offers high-level destructors.
+
+{2:example-1-matching-integer-payload-with-ast_pattern Example 1: Matching Integer Payload with {!Ppxlib.Ast_pattern}}
+
+{{:https://github.com/ocaml-ppx/ppxlib/tree/main/examples/1-AST/b%20-%20Destructing%20AST/destructuring_ast.ml#L37-L40} 🔗 Sample Code}
+
+Let’s destructure the same integer [1] AST using {!Ppxlib.Ast_pattern}:
+
+{[
+open Ppxlib
+
+let match_int_payload =
+  let open Ast_pattern in
+  pstr (pstr_eval (pexp_constant (pconst_integer (string "1") none)) nil ^:: nil)
+]}
+
+This code achieves the same result as the previous example but in a more concise and readable way.
+
+- {b [PStr]} becomes [pstr]
+- {b [Pstr_eval]} becomes [pstr_eval]
+- {b [Pexp_constant]} becomes [pexp_constant]
+- {b [Pconst_integer]} becomes [pconst_integer]
+
+{2:example-2-simplifying-matching-with-eint Example 2: Simplifying Matching with [eint]}
+
+{{:https://github.com/ocaml-ppx/ppxlib/tree/main/examples/1-AST/b%20-%20Destructing%20AST/destructuring_ast.ml#L40-L49} 🔗 Sample Code}
+
+You can further simplify it:
+
+{[
+let match_int_payload =
+  let open Ast_pattern in
+  pstr (pstr_eval (eint (int 1)) nil ^:: nil)
+]}
+
+Using [eint] instead of [pexp_constant] and [pconst_integer] provides better type safety. The [int] wildcard captures the integer value.
```

Comment:
```suggestion
Using [eint] instead of [pexp_constant] and [pconst_integer] provides better type safety. The [int] combinator captures the integer value.
```

I'm not entirely sure about the claim of "better type safety", I think `eint` is just a short-hand for the same code: https://github.com/ocaml-ppx/ppxlib/blob/b3c240e020aba81b7da26ac3c00c2df6710dfda7/src/ast_pattern.ml#L226 ?